sdk/core/azure-core/inc/azure/core/operation.hpp

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Provides support for long-running operations.
+ */
+
+#pragma once
+
+#include "azure/core/context.hpp"
+#include "azure/core/operation_status.hpp"
+#include "azure/core/response.hpp"
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+namespace Azure { namespace Core {
+
+  /**
+   * @brief Methods starting long-running operations return Operation<T> types.
+   *
+   * @tparam T The long-running operation final result type.
+   */
+  template <class T> class Operation {
+  private:
+    // These are pure virtual b/c the derived class must provide an implementation
+    virtual std::unique_ptr<Http::RawResponse> PollInternal(Context& context) = 0;
+    virtual Response<T> PollUntilDoneInternal(Context& context, std::chrono::milliseconds period)
+        = 0;
+
+  protected:
+    OperationStatus m_status = OperationStatus::NotStarted;
+
+  public:
+    /**
+     * @brief Final reuslt of the long-running operation.
+     *
+     * @return Response<T> the final result of the long-running operation.
+     */
+    virtual T Value() const = 0;
```

Comment:
Can we consider something like this so that all implementations of `Value` in the derived types don't have to worry about providing error handling, and we can have a consistent exception type/message?

```C++
    private virtual T ValueImplementation() const = 0;
    T Value() const
    {
      if (!HasValue())
      {
        throw some exception
      }
      return ValueImplementation();
    }
```

Reply:
Think we can also do this for Nullable<>::GetValue()

Reply:
This fits well as a precondition check so it can be turned off in release once the calling code is correctly written to avoid the unnecessary cost of extra checks